cmd/kubeadm/app/cmd/token.go

```diff
@@ -0,0 +1,206 @@
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	"k8s.io/kubernetes/pkg/api"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/kubectl"
+)
+
+const (
+	bootstrapTokenSecretPrefix = "bootstrap-token-"
+)
+
+func NewCmdToken(out io.Writer) *cobra.Command {
+	tokenCmd := &cobra.Command{
+		Use:   "token",
+		Short: "Manage bootstrap tokens",
+	}
+
+	var tokenDuration string
+	createCmd := &cobra.Command{
+		Use:   "create",
+		Short: "Create discovery tokens on the server.",
+		Run: func(tokenCmd *cobra.Command, args []string) {
+			err := RunCreateToken(out, tokenCmd, tokenDuration)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	createCmd.PersistentFlags().StringVar(&tokenDuration,
+		"duration", "8h", "Duration the token should be valid")
+	tokenCmd.AddCommand(createCmd)
+
+	listCmd := &cobra.Command{
+		Use:   "list",
+		Short: "List discovery tokens on the server.",
+		Run: func(tokenCmd *cobra.Command, args []string) {
+			err := RunListTokens(out, tokenCmd)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	tokenCmd.AddCommand(listCmd)
+
+	deleteCmd := &cobra.Command{
+		Use:   "delete",
+		Short: "Delete discovery tokens on the server.",
+		Run: func(tokenCmd *cobra.Command, args []string) {
+			err := RunDeleteToken(out, tokenCmd, args[0])
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	tokenCmd.AddCommand(deleteCmd)
+
+	return tokenCmd
+}
+
+// TODO: Add support for user specified tokens.
+// RunCreateToken generates a new bootstrap token and stores it as a secret on the server.
+func RunCreateToken(out io.Writer, cmd *cobra.Command, tokenDurationStr string) error {
+	client, err := createAPIClient()
+	if err != nil {
+		return err
+	}
+
+	tokenDuration, err := time.ParseDuration(tokenDurationStr)
+	if err != nil {
+		return err
+	}
+
+	tokenSecret := &kubeadmapi.Secrets{}
+	err = kubeadmutil.GenerateToken(tokenSecret)
+	if err != nil {
+		return err
+	}
+
+	secret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name: fmt.Sprintf("%s%s", bootstrapTokenSecretPrefix, tokenSecret.TokenID),
+		},
+		Type: api.SecretTypeBootstrapToken,
+		Data: encodeTokenSecretData(tokenSecret, tokenDuration),
+	}
+	if _, err := client.Secrets(api.NamespaceSystem).Create(secret); err != nil {
+		return fmt.Errorf("<cmd/token> failed to bootstrap token [%v]", err)
+	}
+	fmt.Printf("<cmd/token> Token secret created: %s\n", tokenSecret.GivenToken)
+
+	return nil
+}
+
+func encodeTokenSecretData(tokenSecret *kubeadmapi.Secrets, duration time.Duration) map[string][]byte {
+	var (
+		data = map[string][]byte{}
+	)
+
+	data["token-id"] = []byte(tokenSecret.TokenID)
+	data["token-secret"] = []byte(tokenSecret.BearerToken)
+
+	t := time.Now()
+	t = t.Add(duration)
+	data["expiration"] = []byte(t.Format(time.RFC3339))
+
+	return data
+}
+
+func createAPIClient() (*clientset.Clientset, error) {
```

Comment:
Don't we have some code like this already in `apiclient.go`? If not, we should refactor out this code into a helper